galaxy_ng/app/api/v3/serializers/namespace.py

```diff
@@ -0,0 +1,119 @@
+import logging
+import re
+
+from django.db import transaction
+
+from rest_framework.exceptions import ValidationError
+from rest_framework import serializers
+
+from galaxy_ng.app import models
+from galaxy_ng.app.models import auth as auth_models
+from galaxy_ng.app.api.v3.serializers.group import GroupSummarySerializer
+
+log = logging.getLogger(__name__)
+
+
+class NamespaceLinkSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = models.NamespaceLink
+        fields = ('name', 'url')
+
+
+class NamespaceSerializer(serializers.ModelSerializer):
+    links = NamespaceLinkSerializer(many=True, required=False)
+
+    groups = GroupSummarySerializer(many=True)
+
+    class Meta:
+        model = models.Namespace
+        fields = (
+            'id',
+            'name',
+            'company',
+            'email',
+            'avatar_url',
+            'description',
+            'links',
+            'groups',
+            'resources'
+        )
+
+    # replace with a NamespaceNameSerializer and validate_name() ?
+    def validate_name(self, name):
+        if not name:
+            raise ValidationError(detail={
+                'name': "Attribute 'name' is required"})
+        if not re.match(r'^[a-zA-Z0-9_]+$', name):
+            raise ValidationError(detail={
+                'name': 'Name can only contain [A-Za-z0-9_]'})
+        if len(name) <= 2:
+            raise ValidationError(detail={
+                'name': 'Name must be longer than 2 characters'})
+        if name.startswith('_'):
+            raise ValidationError(detail={
+                'name': "Name cannot begin with '_'"})
+        return name
+
+    @transaction.atomic
+    def create(self, validated_data):
+        log.debug('serializer.create validated_data=%s', validated_data)
+
+        links_data = validated_data.pop('links', [])
+        groups_data = validated_data.pop('groups', None)
+
+        instance = models.Namespace.objects.create(**validated_data)
+
+        ngs = auth_models.Group.objects.filter(name__in=[x['name'] for x in groups_data])
+
+        # create NamespaceLink objects if needed
+        new_links = []
+        for link_data in links_data:
+            ns_link, created = models.NamespaceLink.objects.get_or_create(**link_data)
+            new_links.append(ns_link)
+
+        instance.groups.set(ngs)
+        instance.links.set(new_links)
+
+        return instance
+
+    @transaction.atomic
+    def update(self, instance, validated_data):
```

Comment:
This throws a 500 when the group provided doesn't exist. It would be a little more useful if this threw a validation error.

Reply:
@newswangerd Agreed on the validation error.

But the /v3/_ui/namespaces viewset shouldn't be using this serializer afaik. The ui NamespaceViewset should be using the existing ui.serializers.namespaces.NamespaceUpdate serializer.

Is this with ui talking to the /v3/namespaces/ endpoint and not /v3/_ui/namespaces/ ?

Reply:
I can reproduce like:

```
(.venv_py36) [newswoop:F31:galaxy_ng (namespace_to_v3_180_171 % u=)]$ HTTP_PROXY=localhost:8888 http --session GALAXY_NAMESPACE -v PATCH http://localhost:5001/api/automation-hub/v3/namespaces/testnamespace20/  groups:='[{"id": 1, "name": "system:partner-engineers"}, {"id": 31, "name": "blssdfasdf"}]' links:='[{"name": "CyberspaceOnramp", "url": "http://geocities.com/GlamTown/WorfOrWeen1974"}]'
PATCH /api/automation-hub/v3/namespaces/testnamespace20/ HTTP/1.1


{
    "groups": [
        {
            "id": 1,
            "name": "system:partner-engineers"
        },
        {
            "id": 31,
            "name": "blssdfasdf"
        }
    ],
    "links": [
        {
            "name": "CyberspaceOnramp",
            "url": "http://geocities.com/GlamTown/WorfOrWeen1974"
        }
    ]
}

HTTP/1.1 500 Internal Server Error

<... debug spew ...>
```

Reply:
@newswangerd ok, added a check for exististence of the group in the to_internal_value() to raise a ValidationError in that case.

Not confident that I took the right approach for doing that, but it seems to work.